src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/DatabaseTest.cs

```diff
@@ -11,271 +12,275 @@
-                NullLogger<UpdateDocumentById>.Instance
-            );
-        }
+        // Run the setup
+        await setup(connectionForSetup, transactionForSetup);
 
-        public static GetDocumentById CreateGetById(NpgsqlDataSource dataSource)
-        {
-            return new GetDocumentById(dataSource, new SqlAction(), NullLogger<GetDocumentById>.Instance);
-        }
+        // Commit the setup
+        await transactionForSetup.CommitAsync();
 
-        public static GetDocumentByKey GetDocumentByKey(NpgsqlDataSource dataSource)
-        {
-            return new GetDocumentByKey(dataSource, new SqlAction(), NullLogger<GetDocumentByKey>.Instance);
-        }
+        // Cleanup setup connection/transaction to avoid confusion
+        await transactionForSetup.DisposeAsync();
+        await connectionForSetup.DisposeAsync();
 
-        public static DeleteDocumentById CreateDeleteById(NpgsqlDataSource dataSource)
-        {
-            return new DeleteDocumentById(
-                dataSource,
-                new SqlAction(),
-                NullLogger<DeleteDocumentById>.Instance
-            );
-        }
+        // Connection and transaction managed in this method for DB transaction 1
+        await using var connection1 = await DataSource!.OpenConnectionAsync();
+        await using var transaction1 = await connection1.BeginTransactionAsync(IsolationLevel.Serializable);
 
-        [OneTimeSetUp]
-        public void OneTimeSetup()
-        {
-            new Deploy.DatabaseDeploy().DeployDatabase(_connectionString);
-        }
+        // Use these for threads to signal each other for coordination
+        using EventWaitHandle Transaction1Go = new AutoResetEvent(false);
+        using EventWaitHandle Transaction2Go = new AutoResetEvent(false);
 
-        [SetUp]
-        public async Task SetUp()
-        {
-            DataSource = NpgsqlDataSource.Create(_connectionString);
-            _respawnerConnection = DataSource.OpenConnectionAsync().Result;
-
-            _respawner = await Respawner.CreateAsync(
-                _respawnerConnection,
-                new RespawnerOptions
-                {
-                    TablesToInclude = [new("public", "documents"), new("public", "aliases")],
-                    DbAdapter = DbAdapter.Postgres
-                }
-            );
-        }
+        // Connection and transaction managed in this method for DB transaction 2
+        NpgsqlConnection? connection2 = null;
+        NpgsqlTransaction? transaction2 = null;
+
+        T? result1 = default;
+        U? result2 = default;
 
-        [TearDown]
-        public async Task TearDown()
+        // DB Transaction 2's thread
+        _ = Task.Run(async () =>
         {
-            await _respawner!.ResetAsync(_respawnerConnection!);
-            _respawnerConnection?.Dispose();
-            DataSource?.Dispose();
-        }
+            // Step #0: Wait for signal from transaction 1 thread to start
```

Comment:
Because there's so much back-and-forth between the threads, I numbered what the linear order of operations is.